torch/csrc/lazy/core/trie.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <deque>
+#include <atomic>
+
+#include <c10/core/ScalarType.h>
+
+namespace torch {
+namespace lazy {
+
+class Node;
+using NodePtr = std::shared_ptr<Node>;
+
+struct TORCH_API TrieNode {
+  static int64_t GetNextUniqueId() {
+    static size_t thread_local id_generator = 0;
+    return id_generator++;
+  }
+
+  int64_t unique_id;
+  NodePtr ir_node;
+  std::deque<std::unique_ptr<TrieNode>> successors;
+
+  TrieNode() : unique_id(GetNextUniqueId()), ir_node(nullptr) { }
+  explicit TrieNode(NodePtr node) : unique_id(GetNextUniqueId()), ir_node(std::move(node)) { }
+};
+
+class TORCH_API Trie {
```

Comment:
you should add a docstring to the overall Trie (TraceCache?) class.  It should explain basically how it's used.